### DISCUSSION TITLE
Jira plugin development OSGi vs SPI

### DISCUSSION THREAD

**Discussion by jlahtinen** [Kong/unirest-java#512] [Q&A] [answered]
Any tips how to add Unirest-java as a dependency to jira plugin / app?
Current problem is that when installing app.obr SPI can not find implementation for JsonEngine (https://github.com/Kong/unirest-java/blob/564c3a927bd37c2f224f969fc19bfa51f6e1fbd7/unirest/src/main/java/kong/unirest/core/json/CoreFactory.java#L35). 

My guess is that in osgi world wrong class loader is used. I tried to "fix" this by using spi-fly but couldn't make it work (I guess spy-fly is not included intoOSGi Framework Jira initializes).

**Comment by ryber** [comment]
So I will start off saying that I have no idea how Jira plugins work and am only vaguely familiar with OSGi.

The ServiceLoader is a old school java thing and it works by looking for a special file in the meta-resources of the jars for the Json libraries. like this one: https://github.com/Kong/unirest-java/blob/main/object-mapper-jackson/src/main/resources/META-INF/services/kong.unirest.core.json.JsonEngine

Could it be that the resource file is not present in the however this is getting packaged?

**Reply by jlahtinen** [reply]
For my understanding the file META-INF/services/kong.unirest.core.json.JsonEngine is there correctly included and it contains `kong.unirest.jackson.JacksonEngine` that also looks correct for me.

But SPI still doesn't find any implementations for JsonEngin.

**Comment by ryber** [comment, marked as answer]
Hrm , I have seen issues with ServiceLoader in some packaging systems before, so I've decided to introduce a couple new strategies to CoreFactory:

1. use the java.util.ServiceLoader to load a class by looking for a meta config file in  the resources. They should exist at META-INF.services/kong.unirest.core.json.JsonEngine. The ServiceLoader will use the first one it finds. see https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html

2. It will attempt to load the engine by class name from the classloader by known names in order. These are:
       1. ```kong.unirest.jackson.JacksonEngine```
       2. ```kong.unirest.gson.GsonEngine```
 
3. Clients may set a engine with the setEngine method like ```CoreFactory.setEngine(new JacksonEngine());```

**This is available in Unirest ```4.2.7```**

**Reply by jlahtinen** [reply]
Thanks a lot. I'll test at some point and report back.

**Reply by jlahtinen** [reply]
4.2.7 resolved the original issue.
